File: backend/app/schemas/catalog.py
import uuid
from datetime import date
from enum import StrEnum

from pydantic import BaseModel


class DatasetSort(StrEnum):
    RELEVANCE = "relevance"
    TITLE = "title"
    UPDATED = "updated"
    RECORDS = "records"


class DatasetSummary(BaseModel):
    """Catalog list card — mirrors the prototype's DatasetCard props exactly
    (Master Plan §3 Phase 3 task 1)."""

    id: uuid.UUID
    code: str
    title: str
    category: str | None
    location: str | None
    parameters: list[str]
    source: str | None
    platforms: list[str]
    resolution: str | None
    record_count: int
    formats: list[str]
    license: str | None
    description: str | None
    status: str
    updated_at: str

    model_config = {"from_attributes": True}


class CatalogSearchResponse(BaseModel):
    results: list[DatasetSummary]
    total: int


class SpatialBBox(BaseModel):
    lat_min: float
    lat_max: float
    lon_min: float
    lon_max: float


class DatasetDetail(BaseModel):
    """Full dataset detail — matches the prototype's Dataset type plus
    processing_levels, used by the /catalog/[id] page hero + About panel."""

    id: uuid.UUID
    code: str
    title: str
    category: str | None
    location: str | None
    parameters: list[str]
    source: str | None
    platforms: list[str]
    resolution: str | None
    record_count: int
    formats: list[str]
    processing_levels: list[str]
    license: str | None
    description: str | None
    status: str
    updated_at: str
    temporal_start: date | None
    temporal_end: date | None
    spatial_bbox: SpatialBBox | None

    model_config = {"from_attributes": True}


class StationOption(BaseModel):
    code: str
    name: str
    lat: float
    lon: float


class DatasetRecordPreview(BaseModel):
    """A single previewed row — matches the prototype's Data Preview table
    columns exactly (Date/Station/Depth/Parameter/Value/Unit/Platform/Format/
    Level/Status)."""

    id: uuid.UUID
    time: date
    location: str | None
    depth_m: float | None
    parameter: str
    value: float
    unit: str | None
    platform: str | None
    format: str | None
    processing_level: str | None
    quality_flag: str

    model_config = {"from_attributes": True}


class QualityBreakdown(BaseModel):
    normal: int
    caution: int
    alert: int


class DatasetRecordsResponse(BaseModel):
    """Backs the summary strip (Matching Records / Normal / Caution / Alert /
    % of Dataset) + capped Data Preview table (Master Plan §3 Phase 3 task 3)."""

    preview: list[DatasetRecordPreview]
    matching_count: int
    dataset_total_count: int
    quality_breakdown: QualityBreakdown


class TaxonomyOptions(BaseModel):
    """Populates the catalog filter dropdowns dynamically, replacing the
    prototype's hardcoded taxonomy (Master Plan §3 Phase 3 task 4)."""

    categories: list[str]
    parameters: list[str]
    sources: list[str]
    platforms: list[str]
    formats: list[str]

File: frontend/src/lib/types/catalog.ts
// Mirrors backend/app/schemas/catalog.py exactly — keep these two in sync.

export type DatasetSort = "relevance" | "title" | "updated" | "records";

export interface DatasetSummary {
  id: string;
  code: string;
  title: string;
  category: string | null;
  location: string | null;
  parameters: string[];
  source: string | null;
  platforms: string[];
  resolution: string | null;
  record_count: number;
  formats: string[];
  license: string | null;
  description: string | null;
  status: string;
  updated_at: string;
}

export interface CatalogSearchResponse {
  results: DatasetSummary[];
  total: number;
}

export interface SpatialBBox {
  lat_min: number;
  lat_max: number;
  lon_min: number;
  lon_max: number;
}

export interface DatasetDetail {
  id: string;
  code: string;
  title: string;
  category: string | null;
  location: string | null;
  parameters: string[];
  source: string | null;
  platforms: string[];
  resolution: string | null;
  record_count: number;
  formats: string[];
  processing_levels: string[];
  license: string | null;
  description: string | null;
  status: string;
  updated_at: string;
  temporal_start: string | null;
  temporal_end: string | null;
  spatial_bbox: SpatialBBox | null;
}

export interface StationOption {
  code: string;
  name: string;
  lat: number;
  lon: number;
}

export type QualityFlag = "normal" | "caution" | "alert";

export interface DatasetRecordPreview {
  id: string;
  time: string | null;
  location: string | null;
  depth_m: number | null;
  parameter: string;
  value: number;
  unit: string | null;
  platform: string | null;
  format: string | null;
  processing_level: string | null;
  quality_flag: QualityFlag;
}

export interface QualityBreakdown {
  normal: number;
  caution: number;
  alert: number;
}

export interface DatasetRecordsResponse {
  preview: DatasetRecordPreview[];
  matching_count: number;
  dataset_total_count: number;
  quality_breakdown: QualityBreakdown;
}

export interface TaxonomyOptions {
  categories: string[];
  parameters: string[];
  sources: string[];
  platforms: string[];
  formats: string[];
}

export interface SchemaFilterVariable {
  name: string;
  data_type: string;
  is_dimension: boolean;
  roles: string[];
  min_value: number | null;
  max_value: number | null;
  distinct_values: string[] | null;
}

export interface DatasetSchemaFilters {
  reviewed_at: string;
  variables: SchemaFilterVariable[];
}
